packages/beacon-node/src/network/processor/gossipQueues/indexed.ts

```diff
@@ -0,0 +1,153 @@
+ *   - On next pick the last key with minChunksize
+ *     - if there is no key with minChunkSize, pop the last item of the last key
+ *
+ * This is a special gossip queue for beacon_attestation topic
+ */
+export class IndexedGossipQueue<T extends {indexed?: string}> implements GossipQueue<T> {
+  private _length = 0;
+  private indexedItems: OrderedMap<string, LinkedList<T>>;
+  // keys with at least minChunkSize items
+  // we want to process the last key with minChunkSize first, similar to LIFO
+  private minChunkSizeKeys = new OrderedSet<string>();
+  constructor(private readonly opts: IndexedGossipQueueOpts<T>) {
+    const {minChunkSize, maxChunkSize} = opts;
+    if (minChunkSize < 0 || maxChunkSize < 0 || minChunkSize > maxChunkSize) {
+      throw Error(`Unexpected min chunk size ${minChunkSize}, max chunk size ${maxChunkSize}}`);
+    }
+    this.indexedItems = new OrderedMap<string, LinkedList<T>>();
+  }
+
+  get length(): number {
+    return this._length;
+  }
+
+  get keySize(): number {
+    return this.indexedItems.size();
+  }
+
+  clear(): void {
+    this.indexedItems = new OrderedMap();
+    this._length = 0;
+    this.minChunkSizeKeys = new OrderedSet();
+  }
+
+  /**
+   * Add item to gossip queue. If queue is full, drop first item of first key.
+   * Return number of items dropped
+   */
+  add(item: T): number {
+    const key = this.opts.indexFn(item);
+    if (key == null) {
+      // this comes from getAttDataBase64FromAttestationSerialized() return type
+      // should not happen
+      return 0;
+    }
+    item.indexed = key;
+    let list = this.indexedItems.get(key);
+    if (list == null) {
+      list = new LinkedList<T>();
+      this.indexedItems.set(key, list);
+    }
+    list.push(item);
+    if (list.length >= this.opts.minChunkSize) {
+      this.minChunkSizeKeys.add(key);
+    }
+    this._length++;
+    if (this._length <= this.opts.maxLength) {
+      return 0;
+    }
+
+    // overload, need to drop more items
+    const firstKey = this.indexedItems.firstKey();
+    // there should be at least 1 key
+    if (firstKey == null) {
+      return 0;
+    }
+    const firstList = this.indexedItems.get(firstKey);
+    // should not happen
+    if (firstList == null) {
+      return 0;
+    }
+    const deletedItem = firstList.shift();
+    if (deletedItem != null) {
+      this._length--;
+      if (firstList.length === 0) {
+        // it's faster to search for deleted item from the head in this case
+        this.indexedItems.delete(firstKey, true);
+      }
+      if (firstList.length < this.opts.minChunkSize) {
+        // it's faster to search for deleted item from the head in this case
+        this.minChunkSizeKeys.delete(firstKey, true);
+      }
+      return 1;
+    } else {
+      return 0;
+    }
+  }
+
+  /**
+   * Try to get items of last key with minChunkSize first.
+   * If not, pick the last key
+   */
+  next(): T[] | null {
+    let key: string | null = this.minChunkSizeKeys.last();
```

Comment:
I don't think this approach is best, consider this case:

- a key receives minChunkSize-1 items at time T1 and then 1 extra item at time T2
- It is appended to minChunkSizeKeys at time T2 and will be picked up first, but most of its items are at time T1

Similarly if a key does not receive minChunkSize it will always be de-prioritized. This approach risks respecting FIFO for some load patterns.

The goals for a FIFO queue with indexing that intends to batch should be:
1. Prioritize latest items
2. Buffer items to allow batching, but limit buffering by max wait or max length

For 1. in batches, what decide the timeliness of a batch?
- First arrival
- Latest arrival
- Avg of items

Reply:
yes, this is not ideal. Initially I did like this:
- always pick the last item (like LIFO)
- process it along with `min(64, number of messages with same data/key)`

the performance is far from the current approach since it can't batch much messages. I think it's not easy to pick a strategy, we need to:
- define some criterias to pick a strategy, one is: job wait time + job time + avg(forwarded messages)
- do some dynamic configurations here, configure it at runtime and compare